### PR TITLE
proxy: Ignore visibility annotation if proxy is disabled

### LINKED ISSUE #27594: annotate pod  policy.cilium.io/proxy-visibility causing cilium to panic with WireGuard
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Before 1.14, we can't have l7 proxy with WireGuard. we prevent it from in here https://github.com/cilium/cilium/blob/v1.13/daemon/cmd/daemon_main.go#L1604

However, there is a loophole. if we directly annotate the pod for l7 visibility, that will directly panic cilium. The cilium pod will be in crashloop . I think we should prevent it from panic
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x21b1f90]

goroutine 939 [running]:
github.com/cilium/cilium/pkg/proxy.(*Proxy).CreateOrUpdateRedirect(0x0, {0x31f9e38?, 0x4000f42200}, {0x31fe020, 0x400006f440}, {0x4000efcca8, 0x12}, {0x320fc20, 0x4000f3ce00}, 0x4000191940)
        /go/src/github.com/cilium/cilium/pkg/proxy/proxy.go:491 +0x60
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).addVisibilityRedirects(0x4000f3ce00, 0x0, 0x2c414d0?, 0x6?)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:390 +0x35c
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).addNewRedirects(0x4000f3ce00, 0x4000edd680?)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:476 +0x308
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps(0x4000f3ce00, 0x400137c400, 0x0)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:853 +0x3cc
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0x4000f3ce00, 0x400137c400)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:594 +0x160
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0x4000f3ce00, 0x400137c400)
        /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:457 +0x600
github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle(0x400001e320, 0x40009c0f28?)
        /go/src/github.com/cilium/cilium/pkg/endpoint/events.go:53 +0x2a4
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run.func1()
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:245 +0x108
sync.(*Once).doSlow(0x0?, 0x7e098?)
        /usr/local/go/src/sync/once.go:74 +0x104
sync.(*Once).Do(...)
        /usr/local/go/src/sync/once.go:65
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run(0x40009dcd70?)
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:233 +0x48
created by github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:229 +0x7c
```

### Cilium Version

v1.13.6 v.1.14.1

### Kernel Version

Linux minikube 5.15.0-78-generic #85-Ubuntu SMP Fri Jul 7 15:29:30 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux

### Kubernetes Version

I tested with 1.24.15

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

you can just use the following helm values
```
cluster:
  name: minikube
encryption:
  enabled: true
  nodeEncryption: false
  type: wireguard
endpointStatus:
  enabled: true
  status: policy
l7Proxy: false
operator:
  replicas: 1
serviceAccounts:
  cilium:
    name: cilium
  operator:
    name: cilium-operator
tunnel: vxlan
```
then run 2 nginx pods
```
k run nginx --image=nginx
kubectl annotate pod nginx  policy.cilium.io/proxy-visibility="<Egress/53/UDP/DNS>,<Egress/80/TCP/HTTP>"
k run nginx1 --image=nginx
```
the nginx1 will cause the cilium to crash

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by sayboras**
### Description

This is to avoid the below error if proxy is disabled as part of
installation, but the visibility annotation is added to pods later.

```console
2023-08-11T12:42:41.390575371Z goroutine 1522 [running]:
2023-08-11T12:42:41.390581994Z github.com/cilium/cilium/pkg/proxy.(*Proxy).CreateOrUpdateRedirect(0x0, {0x3ae74b8, 0xc000650280}, {0x3aeb440?, 0xc0016e4560?}, {0xc0010ba1b0, 0x12}, {0x3afd260, 0xc000982000}, 0xc001c9c980)
2023-08-11T12:42:41.390589198Z  github.com/cilium/cilium/pkg/proxy/proxy.go:459 +0xb7
2023-08-11T12:42:41.390596081Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).addVisibilityRedirects(0xc000982000, 0x1, 0xc001874ba0?, 0xc001c9c980?)
2023-08-11T12:42:41.390602613Z  github.com/cilium/cilium/pkg/endpoint/bpf.go:343 +0x443
2023-08-11T12:42:41.390614345Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).addNewRedirects(0xc000982000, 0xc001874870?)
2023-08-11T12:42:41.390651325Z  github.com/cilium/cilium/pkg/endpoint/bpf.go:424 +0x3c5
2023-08-11T12:42:41.390658068Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps(0xc000982000, 0xc000e31800, 0x0)
2023-08-11T12:42:41.390663999Z  github.com/cilium/cilium/pkg/endpoint/bpf.go:802 +0x4fe
2023-08-11T12:42:41.390669690Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0xc000982000, 0xc000e31800)
2023-08-11T12:42:41.390675451Z  github.com/cilium/cilium/pkg/endpoint/bpf.go:542 +0x1a5
2023-08-11T12:42:41.390681112Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0xc000982000, 0xc000e31800)
2023-08-11T12:42:41.390686893Z  github.com/cilium/cilium/pkg/endpoint/policy.go:467 +0x9c6
2023-08-11T12:42:41.390692564Z github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle(0xc000131b50, 0x0?)
2023-08-11T12:42:41.390698385Z  github.com/cilium/cilium/pkg/endpoint/events.go:53 +0x325
2023-08-11T12:42:41.390704045Z github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run.func1()
2023-08-11T12:42:41.390709716Z  github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:245 +0x142
```
    
Fixes: #27594

### Testing

Testing was done locally with l7Proxy=false, and then annotate a pod with visibility annotation

```console
$ k annotate pod client-77bf9989dc-984xp policy.cilium.io/proxy-visibility="<Egress/80/TCP/HTTP>" --overwrite
pod/client-77bf9989dc-984xp annotate
$ ksyslo ds/cilium --timestamps | zgrep "ignoring L7 proxy visibility policy as L7 proxy is disabled"
2023-08-20T13:02:08.241508927Z level=warning msg="ignoring L7 proxy visibility policy as L7 proxy is disabled" ciliumEndpointName=default/client-77bf9989dc-984xp containerID=1ea0a77f5d containerInterface=eth0 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=759 identity=831 ipv4=10.244.0.19 ipv6= k8sPodName=default/client-77bf9989dc-984xp subsys=endpoint
```

### CODE REVIEW

**Review by kaworu — APPROVED** [review]

**Review by youngnick — APPROVED** [review]

File: Documentation/observability/visibility.rst [hunk 1]
```diff
@@ -10,6 +10,10 @@
 Layer 7 Protocol Visibility
 ***************************
 
+.. note::
+
+    This feature requires L7 Proxy support to be enabled. If not, the visibility annotation will be ignored.
```
Comment (zacharysarah) [inline Documentation/observability/visibility.rst]: Use present tense, use active voice
```suggestion
    This feature requires enabling L7 Proxy support. Without it, the visibility annotation is ignored.
```

**Review by zacharysarah — CHANGES_REQUESTED** [review]
@sayboras Small changes for clarity and consistency, otherwise LGTM for docs.

File: Documentation/observability/visibility.rst (hunk 1 above)
Comment (sayboras) [reply Documentation/observability/visibility.rst]: Thanks for your input :+1:

**Review by zacharysarah — APPROVED** [review]

**Review by jrajahalme — APPROVED** [review]

**Pull request merged**